### PR TITLE
Disallow arrow functions

### LINKED ISSUE #1580: Add "pessimistic" arrow function rules
There should be a `disallowArrowFunctions` rule, and a `disallowShorthandArrowFunctions` rule.

It turns out that you don't need `this` to write JavaScript. Object-orientation is perfectly viable with functions that return objects. In fact, it might be better, because if you create objects with functions, you can hide information with closures; whereas everything on `this` has to be publicly accessible.

Since `this` isn't needed, arrow functions' lexical `this` isn't useful. Arrow functions are also less versatile than `function`, because you can't give them names. Arrow functions' tricky syntax is also harmful to maintenance. Adding additional lines to an arrow function which implicitly returns a value will result in that function returning `undefined`, unless you remember to add an explicit `return`. The shorthand syntax is also ambiguous in terms of returning objects, as `(name) => {id: name}` is interpreted as a longhand arrow function with the label `id:`.

Some people might decide to just eliminate them altogether, since `function` seems a lot better if you aren't using `this`. But for people who don't mind `this`, they might want to reap the perceived benefits of arrow functions while avoiding their syntactic problems, by requiring they be written in the most-generalized form, the longhand form. Therefore, it is important these rules be added.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25101917-add-pessimistic-arrow-function-rules?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jacksonrayhamilton**
Fixes #1580.

### CODE REVIEW

**Comment by mrjoelkemp** [comment]
Nice work here @jacksonrayhamilton! Thanks for contributing. 

LGTM

**Comment by jacksonrayhamilton** [comment]
Added justifications per [this suggestion](https://github.com/jscs-dev/node-jscs/issues/1580#issuecomment-127043287).

**Comment by hzoo** [comment]
I'd just like to note we have the `disallowNodeTypes` (http://jscs.info/rule/disallowNodeTypes.html) rule which could be used instead of the `disallowArrowFunctions` rule but doesn't mean we can't have both.

File: lib/rules/disallow-arrow-functions.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * Disallows arrow functions.
+ *
+ * Why enable this rule? Arrow functions might cause more problems than they
+ * solve:
+ *
+ * - Object-orientation may be better without ECMAScript's `this`.
+ * - You can't name an arrow function.
+ * - Arrow functions' syntax can cause maintenance issues; see
+ *   `disallowShorthandArrowFunctions`.
+ * - Arrow functions shouldn't be used on prototypes, as objects' methods,
```
Comment (markelog) [inline lib/rules/disallow-arrow-functions.js]: `objects' `?

File: lib/rules/disallow-shorthand-arrow-functions.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/**
+ * Require arrow functions to use a block statement (explicit return).
+ *
+ * Why enable this rule? Arrow functions' syntax can cause maintenance issues:
```
Comment (markelog) [inline lib/rules/disallow-shorthand-arrow-functions.js]: Same as above

**Comment by markelog** [comment]
Couple notes, otherwise good to go

File: lib/rules/disallow-arrow-functions.js (hunk 1 above)
Comment (ronkorving) [inline lib/rules/disallow-arrow-functions.js]: Possessive form of a plural word ending in 's', seems correct to me.

Comment (mrjoelkemp) [inline lib/rules/disallow-arrow-functions.js]: Yeah, it's correct; though it always looks strange even to native English speakers :)

**Pull request closed without merging**